File: Data Structures & Algorithms/target-sum/submission-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        def backtrack(i, curr_sum):
            if i == len(nums)-1:
                return 1 if curr_sum == target else 0
            if (i, curr_sum) in memo:
                return memo[(i, curr_sum)]

            times = 0
            times += backtrack(i+1, curr_sum+nums[i+1])
            times += backtrack(i+1, curr_sum-nums[i+1])

            memo[(i, curr_sum)] = times
            return times

        memo = {}
        res = 0
        res += backtrack(0, nums[0])
        res += backtrack(0, -nums[0])

        return res

File: Data Structures & Algorithms/binary-search/submission-2.cs
public class Solution {
    public int Search(int[] nums, int target) {
        int low = 0, high = nums.Length - 1;

        while (low <= high){
            int middle = low + (high-low);

            if (nums[middle] == target) return middle;
            if (nums[middle] < target) low = middle + 1;
            else high = middle -1;
        }

        return -1;
    }
}
